### ISSUE TITLE
Implementing a no-copy C++ operator

### ISSUE THREAD

**Issue by fversaci** [NVIDIA/DALI#4172]
Hi,

I've been implementing a C++ data loader, as a CPU-to-CPU DALI operator. It produces in output full batches as `TensorVector`s, which are continuously newly allocated within the data loader and then moved to overwrite `ws.Output<::dali::CPUBackend>(0)`.

My understanding is that these vectors are then copied by the DALI pipeline, which assumes that the allocated space is reused every time. Since this is not my case, I'd like to get rid of this extra copy, to improve the peak performance of the loader. Is there a way to have the DALI pipeline treat these `TensorVector`s as disposable, as it is the case when using an external source from Python, with the `no_copy` parameter set to true?

Thanks!

**Comment by mzient** [comment]
Hello @fversaci, thanks for reaching to us!
Normally, DALI doesn't copy the output of any operator (except CPU->GPU transition). If an operator returns `true` from `CanInferOutputs` and `Setup`, then the Executor will allocate the outputs. If you have some kind of cache which you want to send to the output, you can suppress the allocation of the outputs by returning `false` from these functions. Then you can populate the outputs via a call to `ShareData`. One thing that you must not do is to overwrite the buffer that you've sent to the output in the previous iteration. If the memory needs to be disposed of in a special way (e.g. it's a mapped file), then you need to pass a suitable deleter to `ShareData`.
If hope this sheds some light on the topic. If something remains unclear, don't hesitate to ask further.

**Comment by fversaci** [comment]
Ok, thanks a lot for your answer. I have both `CanInferOutputs` and `Setup` return false.

My original code was something like:
```C++
void MyLoader::RunImpl(::dali::HostWorkspace &ws) {
  ::dali::TensorVector<::dali::CPUBackend> batch = my_get_batch();
  // move features to output
  auto &features = ws.Output<::dali::CPUBackend>(0);
  features = std::move(batch);
}
```
and I have now changed it to:
```C++
void MyLoader::RunImpl(::dali::HostWorkspace &ws) {
  ::dali::TensorVector<::dali::CPUBackend> batch = my_get_batch();
  // move features to output
  auto &features = ws.Output<::dali::CPUBackend>(0);
  // std::cout << features.shape() << std::endl;
  features.ShareData(batch);
}
```
Is that the correct way?

I still have a few doubts:
- If I print (as in the commented line) the shape of the features vector **before** sharing the data I see that, except for the first batch, it is already populated (its shape is not 0,0,...). Doesn't this mean that the vector returned by `pipeline.run()` is obtained by a copy (which I would like to get rid of)?
- If I force the pipeline to copy the data to the GPU (`images = images.gpu()` in the pipeline), the throughput increases, compared to using only the CPU. I would also interpret this as a time saving given by DMA GPU copy vs CPU copy.

**Comment by mzient** [comment]
I'm glad I could be of help.
Regarding your questions:
* the output you see in subsequent calls is what you've set in (one of) the previous iterations
* I haven't seen the entire pipeline, so I cannot really explain why it happens

**Comment by fversaci** [comment]
Thanks again, I'll try and clarify what I mean.

Let's say we have a simple pipeline which just uses `MyLoader` (or also just plain `fn.readers.file`) to read some raw files and returns a `tensor_vector` to the CPU. Let's also assume there is no prefetching (just one buffer). I run the pipeline twice in row:
```python
x = pipeline.run()
y = pipeline.run()
```
As you've said, at the second run the buffer `ws.Output<::dali::CPUBackend>(0)` contains the data from the previous run (i.e., `x`). What happens now if I overwrite this data with something else (no reallocation)? Does that change the values in `x` (i.e., we have `x == y`)?

My guess was that data obtained from different runs (i.e., `x` and `y`) do not share memory space. But if I can see the previous vector each time I run the pipeline I would say that, to keep `x` and `y` independent they have to be copied at some point in time.

Also, even when running a pipeline containing only `fn.readers.file` I can observe an improvement in performance when copying the output to GPU, compared to leaving it in CPU (in my test: 330 vs 310 batch/s).

**Comment by mzient** [comment]
> My guess was that data obtained from different runs (...) do not share memory space

No, It is the same memory. It's generally **not safe** to overwrite its contents*. It is not safe to overwrite a local buffer that has been shared and call ShareData with this buffer again. However, it is **safe** to call ShareData with a different source pointer.

*Until the memory is consumed - which means, until the mixed stage from the corresponding iteration has completed. An operator has no means to ensure this. Also, it's safe to do if you configure your pipeline as `exec_async=False`, but this will adversely affect performance.

Regarding the speedup resulting from CPU->GPU copy - there might be an additional copy at the output of the pipeline - especially if the output is on CPU. Also, if the data is consumed on the GPU, it's better to have the host-to-device copy in DALI than outside.

**Comment by fversaci** [comment]
I understand now, thank you a lot for the detailed explanation!
